pages/controls/Certificate_and_Public_Key_Pinning.md

```diff
@@ -357,40 +357,142 @@ However, implementing pinning validation from scratch should be avoided,
 as implementation mistakes are extremely likely and usually lead to
 severe vulnerabilities.
 
-### .Net
+### .NET
 
-.Net pinning can be achieved by using `ServicePointManager` as shown
-below.
+.NET pinning can be achieved by using `ServicePointManager` for `HttpWebRequest`, or 
+`HttpClientHandler` when using `HttpClient`, as shown below.
+
+#### HttpWebRequest
 
 Download: [.Net sample program](https://wiki.owasp.org/images/2/25/Pubkey-pin-dotnet.zip).
 
 ```
 // Encoded RSAPublicKey
-private static String PUB_KEY = "30818902818100C4A06B7B52F8D17DC1CCB47362" +
-    "C64AB799AAE19E245A7559E9CEEC7D8AA4DF07CB0B21FDFD763C63A313A668FE9D764E" +
-    "D913C51A676788DB62AF624F422C2F112C1316922AA5D37823CD9F43D1FC54513D14B2" +
-    "9E36991F08A042C42EAAEEE5FE8E2CB10167174A359CEBF6FACC2C9CA933AD403137EE" +
-    "2C3F4CBED9460129C72B0203010001";
+private static string PUB_KEY = "30818902818100C4A06B7B52F8D17DC1CCB47362" +
+	"C64AB799AAE19E245A7559E9CEEC7D8AA4DF07CB0B21FDFD763C63A313A668FE9D764E" +
+	"D913C51A676788DB62AF624F422C2F112C1316922AA5D37823CD9F43D1FC54513D14B2" +
+	"9E36991F08A042C42EAAEEE5FE8E2CB10167174A359CEBF6FACC2C9CA933AD403137EE" +
+	"2C3F4CBED9460129C72B0203010001";
+
+public static void Main(string[] args)
+{
+	ServicePointManager.ServerCertificateValidationCallback = PinPublicKey;
+	var wr = WebRequest.Create("https://encrypted.google.com/");
+	wr.GetResponse();
+}
+
+public static bool PinPublicKey(object sender, X509Certificate certificate, X509Chain chain,
+								SslPolicyErrors sslPolicyErrors)
+{
+	// No errors
+	if (sslPolicyErrors == SslPolicyErrors.None)
+		return true;
```

Comment:
Like other comment, won't this skip the below checks?

Reply:
With `ServicePointManager.ServerCertificateValidationCallback`, the `PinPublicKey` method will be called for _every_ HTTP request from the application. 

In the example, only a request to `https://encrypted.google.com/` is made, but in a real application, it's likely that requests are made to multiple different endpoints. If a request is made to a HTTPS endpoint that presents a valid certificate ([`SslPolicyErrors.None`](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslpolicyerrors?view=net-5.0)), then these should be allowed to succeed and not fall through to certificate pinning logic that applies only to `encrypted.google.com`. For any other HTTPS request that has policy errors that isn't `encrypted.google.com`, validation will fail.

The certificate pinning logic would be skipped for a HTTP request to `encrypted.google.com` _only_ if it presented a valid certificate. In such a scenario, one would question the need to apply certificate pinning, if the certificate for the endpoint for which certificate pinning is applied is already valid 🙂 

I'll defer the decision on whether to keep this check in to your good selves. If the intention is to provide an example that a developer can copy/paste, and just change the key string, I think it would be useful to keep it in.

Reply:
Ah, I get what you're trying to do with that.

The problem is, there are a number of threats that certificate pinning helps mitigate that involves not trusting the usual certificate authority trust chain.

A compromised CA that allows an attacker to issue their own "valid" certs https://threatpost.com/final-report-diginotar-hack-shows-total-compromise-ca-servers-103112/77170/

Gaining control of a DNS zone or compromising a web server would also allow you to use a CA like Let's Encrypt to issue "valid" certs.

An unintentional bug at one of the CAs https://www.zdnet.com/article/lets-encrypt-to-revoke-3-million-certificates-on-march-4-due-to-bug/

I think any included example needs to default to strict pinning and ignore CA trust chains. Unless the example is demonstrating pinning a particular part of the trust chain like an intermediary cert.

Reply:
Good points. Perhaps checking the hostname would also be useful, something like

```suggestion
        var request = sender as HttpWebRequest;
	if (request == null)
	        return false;

	// A valid certificate for another hostname was presented
	if (sslPolicyErrors == SslPolicyErrors.None && !string.Equals(request.Address.Authority, "encrypted.google.com"))
		return true;
```

Whether to include this, I think comes down to the intention behind the example; if it's intended to show pinning for a specific hostname in isolation, removing it would simplify the example, but if it's intended to show pinning for a specific hostname where requests might be made to other hosts, it might be useful to keep in. Have you had feedback on examples that could help guide a decision?

Reply:
I think, given the context of how HttpClient should be used in .NET applications, that is a really good example. It covers strict pinning but also shows how to handle requests to hosts you don't want to pin. And also keeps it simple for people to tailor it to their own requirements.